cpp/include/raft/cluster/detail/kmeans.cuh

```diff
@@ -19,967 +19,1056 @@
+  rmm::device_uvector<DataT> L2NormBuf_OR_DistBuf(0, stream);
+
+  // temporary buffer to store intermediate centroids, destructor releases the
+  // resource
+  auto newCentroids = raft::make_device_matrix<DataT>(n_clusters, n_features, stream);
+
+  // temporary buffer to store weights per cluster, destructor releases the
+  // resource
+  auto wtInCluster = raft::make_device_vector<DataT>(n_clusters, stream);
+
+  rmm::device_scalar<cub::KeyValuePair<IndexT, DataT>> clusterCostD(stream);
+
+  // L2 norm of X: ||x||^2
+  auto L2NormX = raft::make_device_vector<DataT>(n_samples, stream);
+  if (metric == raft::distance::DistanceType::L2Expanded ||
+      metric == raft::distance::DistanceType::L2SqrtExpanded) {
+    raft::linalg::rowNorm(
+      L2NormX.data(), X.data(), X.extent(1), X.extent(0), raft::linalg::L2Norm, true, stream);
   }
-}
 
-/**
- *  @brief Divide rows of centroid matrix by cluster sizes.
- *    Divides the ith column of the sum matrix by the size of the ith
- *    cluster. If the sum matrix has been initialized so that the ith
- *    row is the sum of all observation vectors in the ith cluster,
- *    this kernel produces cluster centroids. The grid and block
- *    dimensions should be 2-dimensional. Ideally the grid is large
- *    enough so there are d threads in the x-direction and k threads
- *    in the y-direction.
- *  @tparam index_type_t the type of data used for indexing.
- *  @tparam value_type_t the type of data used for weights, distances.
- *  @param d Dimension of observation vectors.
- *  @param k Number of clusters.
- *  @param clusterSizes (Input, k entries) Number of points in each
- *    cluster.
- *  @param centroids (Input/output, d*k entries) Sum matrix. Matrix
- *    is stored column-major and matrix dimensions are d x k. The
- *    ith column is the sum of all observation vectors in the ith
- *    cluster. On exit, the matrix is the centroid matrix (each
- *    column is the mean position of a cluster).
- */
-template <typename index_type_t, typename value_type_t>
-static __global__ void divideCentroids(index_type_t d,
-                                       index_type_t k,
-                                       const index_type_t* __restrict__ clusterSizes,
-                                       value_type_t* __restrict__ centroids)
-{
-  // Global indices
-  index_type_t gidx, gidy;
-
-  // Current cluster size
-  index_type_t clusterSize_private;
-
-  // Observation vector is determined by global y-index
-  gidy = threadIdx.y + blockIdx.y * blockDim.y;
-  while (gidy < k) {
-    // Get cluster size from global memory
-    clusterSize_private = clusterSizes[gidy];
-
-    // Add vector entries to centroid matrix
-    //   vector entris are determined by global x-index
-    gidx = threadIdx.x + blockIdx.x * blockDim.x;
-    while (gidx < d) {
-      centroids[IDX(gidx, gidy, d)] /= clusterSize_private;
-      gidx += blockDim.x * gridDim.x;
+  /*RAFT_LOG_INFO(
```

Comment:
Any commented out log statements should be using `RAFT_LOG_DEBUG` or `RAFT_LOG_TRACE`. The purpose of having different logging levels is to remove the need for commenting them out in the code.